s1m0nWasTaken/CS4337_BigDataProject
# AuthService using Google OAuth 2.0 and JWT

## Summary
[Trello task](https://trello.com/c/ekWZlLs4)
Set up authentication using Google OAuth and role based authorization using JWT and refresh tokens.
Created an **Auth DB** with an **auth** table to store refresh tokens associated with a user ID.

The following link can be used to **login** with Google OAuth: [Login with Google](https://accounts.google.com/o/oauth2/v2/auth?redirect_uri=http://localhost:8081/grantcode&response_type=code&client_id=529138320852-h26t99u2jh694u7q3u3c2oaqma07oabe.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid&access_type=offline)
Then we get an **authorization code** from Google and are redirected to the **_grantcode_** endpoint which uses this code to get an **access token** from Google. We use this access token to get **profile details** from Google.
Then we check if a user exists by that email and if so we login, otherwise the user needs to be registered.
Upon successful login/registration **a JWT and refresh token** is issued.

Only refresh tokens need to be stored in a DB since they will need to be checked when a request is sent to refresh tokens.

The **_refresh_** endpoint should only be called when the _access token expires_ (this can be checked without calling AuthService) and then it checks that the refresh token is _valid and not expired_ and if so returns a **new access token and refresh token**.

Access tokens are valid for 1 hour and refresh tokens are valid for 10 days (can be changed if needed).

Some things to note:
- I had to make the AuthService and UserService use the same external network in docker-compose.yml.
- I am using a UUID generator to make refresh tokens. There might be a better way to do this.
- I had to add another endpoint to UserService to get a user by their email _user/email/{email}_
- I made a random secret key for JWTs in Base64.
- I copied the User model and UserType into AuthService, is there a better way to do this without duplicating code?

## How to verify
- First you will need to put in the correct values for the client secret and secret key as these shouldn't be committed.
![image](https://github.com/user-attachments/assets/321ae266-9957-48de-b0f3-0eed170ff7a4)
_env.py_
- Then you need to make sure you are a test user for our Google OAuth, I can add your email to the test users to do this.
- Build and start the AuthService and UserService.
- Then you can enter the [login link](https://accounts.google.com/o/oauth2/v2/auth?redirect_uri=http://localhost:8082/grantcode&response_type=code&client_id=529138320852-h26t99u2jh694u7q3u3c2oaqma07oabe.apps.googleusercontent.com&scope=https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.email+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile+openid&access_type=offline) in your browser and login via Google OAuth. The response should contain the access (JWT) and refresh token.
- Then in Postman you can try hit the refresh token by hitting /refresh with the refresh token in the header.
- The response should contain a new access and refresh token.
![image](https://github.com/user-attachments/assets/3c50c335-f5c1-46a2-a337-8733eeef238d)

- The JWT can be validated and roles can be extracted from it. This can be done separately in each microservice but will require basically copy and pasting classes. I am looking into the best way to do this and will make a separate PR for it.

## Tests
No tests.

## Related
[Still need to look at how each service doesn't have to implement JWT utilities separately and avoid the duplication of code.](https://trello.com/c/YoXz35jf)
